Ansible refresh inventory credentials

durandom opened the pull request:
# Refreshes credentials for Ansible Tower

* I only include `:name` and `:userid` for now, because I dont know how the other fields should map to our model

* I dont like the massive sub classing of credentials - isnt there a field in `authentications` which we can just re-use and fill from ansible? Now we have to touch this code whenever ansible introduces an new credentials type

* The naming of `authentication_configuration_script_bases` caused some severe pain in my head. Can we try to give it a bit more descriptive name? Like `ConfigurationScriptRelationToAuthentication`? Not really much better, but still.

@miq-bot assign @blomquisg 
@Ladas please review the inventory refresh part
@jameswnl please review the credentials part

Ladas commented on app/models/manageiq/providers/automation_manager.rb:
```diff
@@ -5,6 +5,9 @@ class ManageIQ::Providers::AutomationManager < ::ExtManagementSystem
   has_many :configured_systems,           :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_profiles,       :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_scripts,        :dependent => :destroy, :foreign_key => "manager_id"
+  has_many :authentication_configuration_script_bases, :through => :configuration_scripts
```
hm I see configuration_scripts model should also define authentication_configuration_script_bases and ::dependent => :destroy

otherwise this will result in zombies when configuration_script gets deleted

Ladas replied:
in general, this applies to all relations, we usually need :dependent => :destroy or :nullify everywhere

durandom commented on app/models/manageiq/providers/automation_manager.rb:
```diff
@@ -5,6 +5,7 @@ class ManageIQ::Providers::AutomationManager < ::ExtManagementSystem
   has_many :configured_systems,           :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_profiles,       :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_scripts,        :dependent => :destroy, :foreign_key => "manager_id"
+  has_many :configuration_script_authentications, :through => :configuration_scripts, :source => :authentications
```
@Ladas I dont think we'll need `:dependent => :destroy` here, right?
Because its already [here](https://github.com/durandom/manageiq/blob/5287aaf56c29d62c4d989b12553716ccd33c4720/app/models/configuration_script_base.rb#L14-L14)

Ladas replied:
yes, I was looking at a bad model, it's all great :+1:

jameswnl replied:
This relation is there in `ExtManagementSystem` through the inclusion of `AuthenticationMixin`

durandom replied:
> This relation is there in ExtManagementSystem through the inclusion of AuthenticationMixin

sorry, what do you mean exactly? I dont see `configuration_script_authentications` there..

jameswnl replied:
So maybe I don't understand the need here:

Is this meant to get all the `Authentication` this manager has? 
Or `Authentications` that are being used with some `ConfigurationScript`s?

mkanoor replied:
@durandom @jameswnl @bdunne 
Shouldn't all the credentials be connected to a specific EMS, a configuration script (job template) might refer the credential but a configuration script doesn't own the credentials.
From a ems we should be able to call ems.authentications to get all the related credentials.

durandom replied:
@mkanoor thanks for the input, I'll update this.

Right now the credentials are shared across `configuration_scripts` aka `job_templates`. This is what `configuration_script_authentications` links to. I might get away without this if I link all authentications to `ems` *and* share them

bdunne replied:
I would expect `Authentication#resource` to be the owning Manager or Provider and the `has_many :authentications, :dependent => :destroy` should use that.

bdunne commented on app/models/manageiq/providers/ansible_tower/inventory/parser/automation_manager.rb:
```diff
@@ -43,10 +51,51 @@ def configuration_script_sources
+
+  def credentials
+    collector.credentials.each do |i|
+      o = target.configuration_script_authentications.find_or_build(i.id.to_s)
```
I believe it's `AuthenticationConfigurationScriptBases` unless this is something different.

Ladas replied:
this seems correct, AuthenticationConfigurationScriptBases is the M:N mapping table

Ladas commented:
@bdunne @jameswnl we need configuration_script_authentications for refresh, which returns just authentications we got from the API. Using ems.authentications in refresh would delete Ansible auth we added manually?

durandom commented on app/models/manageiq/providers/automation_manager.rb:
```diff
@@ -5,6 +5,8 @@ class ManageIQ::Providers::AutomationManager < ::ExtManagementSystem
   has_many :configured_systems,           :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_profiles,       :dependent => :destroy, :foreign_key => "manager_id"
   has_many :configuration_scripts,        :dependent => :destroy, :foreign_key => "manager_id"
+  has_many :credentials,                  :class_name => "ManageIQ::Providers::AutomationManager::Authentication",
```
How about renaming `ManageIQ::Providers::AutomationManager::Authentication` to `ManageIQ::Providers::AutomationManager::Credential` ?

Ladas replied:
sounds good, for cloud we do rename it to KeyPair

bdunne replied:
In the Ansible Tower namespace I'm fine with using Ansible Tower terms (Credential), but in the base AutomationManager I think it would be less confusing to stick with Authentication

durandom replied:
Tbh, I think `ManageIQ::Providers::AutomationManager::Credential`  less confusing, because it does not conflate the term `Authentication` with those connected to a Manager via `has_many :authentications` - that is those `Authentications` that are used via `Endpoints` to connect to the manager.

It makes it clear, hey here is another concept.

Plus it is more aligned with the association name.

blomquisg replied:
Long term:  I'd like to see two different tables and models (naming is hard, so I don't really care about the specific names used in my list below):

1. `Authentications` for storing authentication data used by MIQ for authenticating on the users' behalf
2. `Credentials` for storing provider inventory records representing credentials that are used by the provider in transactions.

These `AutomationManger` authentication records fall into the second category.

I don't really care which name we use now for this association.  But, I want there to be a clear path forward for distinguishing these types of things in the future so we can refactor it later.

jameswnl replied:
so why do we need this relationship definition here again? Isn't it already there in `AuthenticationMixin` [here](https://github.com/jameswnl/manageiq/blob/ab40ebedc07c7ebd06ac4d9f759cafada5bbfdfb/app/models/mixins/authentication_mixin.rb#L5-L5) and `EMS` included that mixin?

durandom replied:
because it's different. 

[here](https://github.com/ManageIQ/manageiq/blob/master/app/models/manageiq/providers/ansible_tower/automation_manager.rb#L9) we delegate `:authentications` to the provider. Because we expect `ems.authentications` to be those that are 1) in @blomquisg list.

This association here is for 2) in his list.

Thats why I'm arguing for a rename

jameswnl replied:
Didn't see Greg's list when typing up that comment. I am all for it and I've been raising that up from time to time (and stop short before people got annoyed 😅 )

mkanoor commented:
@durandom 
Can you please expose this method via the service_model
https://github.com/ManageIQ/manageiq/blob/40122002df38f24c37ad6dd1657b07696c4dc747/lib/miq_automation_engine/service_models/miq_ae_service_manageiq-providers-ansible_tower-automation_manager.rb

bdunne commented on app/models/manageiq/providers/ansible_tower/inventory/collector/automation_manager.rb:
```diff
@@ -18,4 +18,8 @@ def job_templates
   end
+
+  def credentials
+    @connection.api.credentials.all
```
Shouldn't this just be `connection`, no `@`?

durandom replied:
yes, you are right

durandom commented:
https://www.pivotaltracker.com/story/show/137444851

bdunne commented on app/models/manageiq/providers/ansible_tower/inventory/parser/automation_manager.rb:
```diff
@@ -31,6 +32,14 @@ def configuration_scripts
+
+      o[:authentications] = []
+      %w(credential_id cloud_credential_id network_credential_id).each do |credential_attr|
+        next unless i.respond_to?(credential_attr)
```
Is `i` an `AnsibleTowerClient::JobTemplate`?

durandom replied:
yes, it is.
I can rename it `job_template` - I dont like one letter vars anyway, so dont know what drove me there

durandom commented:
guys, thanks for the review.

I added a commit to address the one letter variables
I also renamed `Authentication` to `Credential`

@bdunne @blomquisg ?

durandom commented:
> I also renamed Authentication to Credential

I remove that commit again, as it also involves some refactoring in automate.
I'll put up PR on that later once this is merged.

jameswnl commented on app/models/manageiq/providers/ansible_tower/inventory/parser/automation_manager.rb:
```diff
@@ -4,49 +4,99 @@ def parse
+                                when 'net' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::NetworkCredential'
+                                when 'ssh' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::MachineCredential'
+                                when 'vmware' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::VmwareCredential'
+                                # when 'scm' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::???Credential'
```
This `ScmCredential` is coming in #13776

jameswnl commented on app/models/manageiq/providers/ansible_tower/inventory/parser/automation_manager.rb:
```diff
@@ -4,49 +4,99 @@ def parse
+                                when 'aws' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::AmazonCredential'
+                                when 'rax' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::RackspaceCredential'
+                                when 'satellite6' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::Satellite6Credential'
+                                # when 'cloudforms' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::$$$Credential'
```
We don't likely to manage CF from CF through Tower and so this is not modeled.

Ladas replied:
Hm, but when tower is set to get inventory from CF, we will need this right?

jameswnl commented on app/models/manageiq/providers/ansible_tower/inventory/parser/automation_manager.rb:
```diff
@@ -4,49 +4,99 @@ def parse
+                                when 'satellite6' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::Satellite6Credential'
+                                # when 'cloudforms' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::$$$Credential'
+                                when 'gce' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::GoogleCredential'
+                                when 'azure' then 'ManageIQ::Providers::AnsibleTower::AutomationManager::AzureCredential'
```
This class actually corresponds to `azure_rm` type as stated [here](https://github.com/jameswnl/manageiq/blob/440a6dc19bd8b97e55abc1aea67d72d30bbf7a0e/app/models/manageiq/providers/ansible_tower/automation_manager/azure_credential.rb#L1-L1).

durandom replied:
Ha, I was thinking, `azure_rm` is deprecated. As in Azure Remove...

durandom commented:
Changed to the correct azure credential. So I guess it's good to go.

@blomquisg mergie merge?
